added XmpDirectory.Root and classes for writing Xmp to Jpeg files

michaelosthege opened the pull request:
Adds a Root property to XmpDirectory such that the XDocument can be accessed directly.
Load behavior is modified to set the Root property.
Writing is accomplished by substituting existing Xml in App1 segments with new Xml, or inserting it if not present in the first place.

See #23 for more information.

drewnoakes commented:
Just wanted to say that I'm pumped to get this feature into the library, but I want to give it a proper review because it'll be the first time the library is writing data and therefore the first time the library could potentially destroy users' files. I'm really busy at the moment but hope to find some time for this soon. Thanks for your patience.

drewnoakes commented:
Quick update. I've reworked the JPEG _reading_ code to be more robust. It will allow better performance and safety for injecting/replacing XMP in JPEG files.  Haven't pushed it yet as I want to do some more tidying and it's late here now.

michaelosthege commented:
I just updated my branch with your latest changes & also some refactorings.
+ refactored it such that the algorithm is easier to understand, review & test,
+ Unit tests for the Jpeg Xmp writing are still ToDo - any suggestions for a good test case?

I had used `http://ns.adobe.com/xap/1.0/\x0` as a segment preamble before and now I use `XmpReder.JpegSegmentPreamble` wich equals `http://ns.adobe.com/xap/1.0/\0` (without the x).  (My reference was [this](http://dev.exiv2.org/projects/exiv2/wiki/The_Metadata_in_JPEG_files))

In both reading & [writing](https://github.com/drewnoakes/metadata-extractor-dotnet/pull/65/files#diff-cf437b89164f79344877a2490237b79fR154), **could the differences in the end of the preamble lead to segments that are not recognized as XMP segments?** (And thereby ignoring them in read, or adding a duplicate segment in write?)

elangelo commented:
@drewnoakes any chance this can be picked up again? i'm willing to merge the conflicts and to follow up on any remarks you still have after review.

drewnoakes commented:
@elangelo the barriers to merging this are 1) my free time to review and really think deeply about this major change to the library and 2) my reluctance to take on the risk and responsibility involved with writing metadata to files and potentially breaking people's data. The cost of bugs becomes _much_ higher once we start writing data. That said, there's regular interest in doing so. If you're interested in helping with this PR, then reviewing the general approach and thinking through how other kinds of metadata may be written would be helpful. Thanks.
